src/vehicle.cpp

```diff
@@ -2643,12 +2673,114 @@ void vehicle::center_of_mass(int &x, int &y)
     y = int(yf + 0.5);
 }
 
+bool vehicle::can_hold_type( int p, const ammotype &ftype )
+{
+    const vehicle_part &tank = parts[p];
+    const vpart_info &tank_info = part_info( p );
+    return tank_info.fuel_type == ftype ||
+        ( tank_info.fuel_type.empty() &&
+            ( tank.items.empty() ||
+              tank.items.front().typeId() == ftype ) );
+}
+
+int vehicle::tank_charges( int p )
+{
+    const vehicle_part &tank = parts[p];
+    if( !tank.items.empty() ) {
+        return tank.items.front().charges;
+    } else {
+        return 0;
+    }
+}
+
+int vehicle::capacity_left( int p, const ammotype &ftype )
+{
+    const vehicle_part &tank = parts[p];
+    const vpart_info &tank_info = part_info( p );
+    if( ftype.empty() ) {
+        int occ = tank.items.empty() ? 0 : tank.items.front().volume();
+        return tank_info.size - occ;
+    }
+    // Incompatible fuel types
+    if( ( !tank_info.fuel_type.empty() && tank_info.fuel_type != ftype ) ||
+        ( !tank.items.empty() && tank.items.front().typeId() != ftype ) ) {
+        return 0;
+    }
+    const itype *type = item_controller->find_template( ftype );
+    const int mul = type->ammo.get() == nullptr ? 1 : type->ammo->def_charges;
+    if( tank.items.empty() ) {
+        return tank_info.size * mul;
+    }
+    return ( tank_info.size - tank.items.front().volume() ) * mul;
+}
+
+void vehicle::tank_set_charges( int p, const ammotype &ftype, int charges )
+{
+    vehicle_part &tank = parts[p];
+    const vpart_info &tank_info = part_info( p );
+    if( tank_info.fuel_type != ftype && charges > 0 ) {
+        debugmsg( "vehicle::tank_set_charges tried to fill %s tank with %s", part_info( p ).fuel_type.c_str(), ftype.c_str() );
+        return;
+    }
+    tank.items.clear();
+    if( charges > 0 ) {
+        item fuel( ftype, calendar::turn );
+        fuel.charges = charges;
+        tank.items.push_front( fuel );
+    }
+}
+
+int vehicle::tank_drain( int p, int charges )
+{
+    vehicle_part &tank = parts[p];
+    if( !tank.items.empty() ) {
+        if( tank.items.front().charges > charges ) {
+            tank.items.front().charges -= charges;
+            return charges;
+        }
+
+        charges -= tank.items.front().charges;
+        tank.items.pop_front();
+        return charges;
+    }
+
+    return 0;
```

Comment:
What is this function supposed to return? The number of charges that have been removed (line 2739 and the `return 0` indicate this) or the number of charges that are still to be removed (line 2744 indicates this).